src/protobuf.cpp

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright The async-profiler authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "protobuf.h"
+#include <string.h>
+
+ProtobufBuffer::~ProtobufBuffer() {
+  if (_parent_message == nullptr) {
+    return;
+  }
+  _parent_message->commitMessage(offset());
+}
+
+void ProtobufBuffer::tag(protobuf_index_t index, protobuf_t type) {
+  putVarInt<>((u32) index << 3 | type);
+}
+
+void ProtobufBuffer::field(protobuf_index_t index, bool b) {
+  field(index, (u32)b);
+}
+
+void ProtobufBuffer::field(protobuf_index_t index, u32 n) {
+  tag(index, VARINT);
+  putVarInt<>(n);
+}
+
+void ProtobufBuffer::field(protobuf_index_t index, u64 n) {
+  tag(index, VARINT);
+  putVarInt<>(n);
+}
+
+void ProtobufBuffer::field(protobuf_index_t index, const char *s) {
+  field(index, s, strlen(s));
+}
+
+void ProtobufBuffer::field(protobuf_index_t index, const char *s, size_t len) {
+  tag(index, LEN);
+  putVarInt<>(len);
+  memcpy(_data + _offset, s, len);
```

Comment:
Do you check that message fits the buffer or do you automatically grow the buffer?

Reply:
That's something I wanted to discuss, I'm not sure if this should be a smarter buffer like `Proto.java`, or a dumb one like what we have in JFR. For now, it's dumb and it gets a `char*` from the user, but I'm open to feedback

Reply:
I'm trying to think of a scenario in which it makes sense not to have it smart enough to resize itself, but I can't find any. I'll make it self-resizable